### PR TITLE
Test and suite creation

### PR DESCRIPTION
**Pull request by gJubey**

### CODE REVIEW

File: src/gui/VerificationValidationWidget.cpp [hunk 1]
```diff
@@ -37,13 +45,51 @@ VerificationValidationWidget::~VerificationValidationWidget() {
...
+    setupUI();
     selectTestsDialog->exec();
     //connect(selectTestsDialog, SIGNAL(accepted()), this, SLOT(runTests()));
 }
 
+QString VerificationValidationWidget::constructTestCommand(TestItem item){
+    QString cmd = item.testCommand;
+    QSqlQuery* q = new QSqlQuery(getDatabase());
+    
+    q->prepare("Select arg, defaultVal from TestArg Where testID = :id ORDER by argIdx");
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:74]: SQL commands have inconsistent format, all caps would help with readability (e.g. SELECT)

File: include/VerificationValidationWidget.h [hunk 2]
```diff
@@ -28,14 +30,36 @@
...
 public:
     explicit VerificationValidationWidget(MainWindow *mainWindow, Document *document, QWidget *parent = nullptr);
     ~VerificationValidationWidget();
     void showSelectTests();
+    void showNewTestDialog();
+    void showNewTestSuiteDialog();
+    void showRemoveTestSuiteDialog();
+    void showRemoveTestDialog();
+    QString* runTest(const QString& cmd);
+    //void runTests();
```
Comment (daperego) [inline include/VerificationValidationWidget.h:66]: Duplicate functions, maybe go back and review which is correct

File: src/gui/MainWindow.cpp [hunk 3]
```diff
@@ -605,6 +605,46 @@ void MainWindow::prepareUi() {
     verifyValidateMenu->addAction(verifyValidateViewportAct);
 
 
+    QAction* verificationValidationNewTest = new QAction(tr("Create new test"), this);
+    verificationValidationNewTest->setIcon(QPixmap::fromImage(coloredIcon(":/icons/verifyValidateIcon.png", "$Color-MenuIconVerifyValidate")));
+    verificationValidationNewTest->setStatusTip(tr("Create new test"));
+    connect(verificationValidationNewTest, &QAction::triggered, this, [this](){
+        if (activeDocumentId == -1) return;
+        documents[activeDocumentId]->getVerificationValidationWidget()->setStatusBar(statusBar);
```
Comment (daperego) [inline src/gui/MainWindow.cpp]: Remove status bar set, from all connections

File: src/gui/VerificationValidationWidget.cpp [hunk 4]
```diff
@@ -7,6 +7,15 @@ using Parser = VerificationValidation::Parser;
 
 #define SHOW_ERROR_POPUP true
 
+TestItem::TestItem(QListWidgetItem* _testWidgetItem, int _id, QString _testName, QString _testCommand, bool _hasValArgs, QString _category){
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Move these arguments to the Test class in VerificationValidation.h

File: src/gui/VerificationValidationWidget.cpp [hunk 5]
```diff
@@ -37,13 +45,51 @@ VerificationValidationWidget::~VerificationValidationWidget() {
...
+}
 
 void VerificationValidationWidget::showSelectTests() {
     statusBar->showMessage("Select tests to run...");
+    setupUI();
     selectTestsDialog->exec();
     //connect(selectTestsDialog, SIGNAL(accepted()), this, SLOT(runTests()));
 }
 
+QString VerificationValidationWidget::constructTestCommand(TestItem item){
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:70]: TestItem will be changing to Test

File: src/gui/VerificationValidationWidget.cpp [hunk 6]
```diff
@@ -37,13 +45,51 @@ VerificationValidationWidget::~VerificationValidationWidget() {
...
+    QSqlQuery* q = new QSqlQuery(getDatabase());
+    
+    q->prepare("Select arg, defaultVal from TestArg Where testID = :id ORDER by argIdx");
+    q->bindValue(":id", item.id);
+    dbExec(q);
+
+    while(q->next()){
+        cmd = cmd + " " + q->value(0).toString();
+        if(q->value(1).toString() != NULL){
+            cmd = cmd + " " + q->value(1).toString();
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Remove " "

File: src/gui/VerificationValidationWidget.cpp [hunk 7]
```diff
@@ -209,7 +255,8 @@ void VerificationValidationWidget::dbInitTables() {
     if (!getDatabase().tables().contains("Model"))
         delete dbExec("CREATE TABLE Model (id INTEGER PRIMARY KEY, filepath TEXT NOT NULL UNIQUE, md5Checksum TEXT NOT NULL)");
     if (!getDatabase().tables().contains("Tests"))
-        delete dbExec("CREATE TABLE Tests (id INTEGER PRIMARY KEY, testName TEXT NOT NULL, testCommand TEXT NOT NULL UNIQUE)");
+        dbExec("CREATE TABLE Tests (id INTEGER PRIMARY KEY, testName TEXT NOT NULL, testCommand TEXT NOT NULL, hasValArgs BOOL NOT NULL, category TEXT NOT NULL)");
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Add delete in front of dbExec()

File: src/gui/VerificationValidationWidget.cpp [hunk 8]
```diff
@@ -220,8 +267,8 @@ void VerificationValidationWidget::dbInitTables() {
         delete dbExec("CREATE TABLE TestSuites (id INTEGER PRIMARY KEY, suiteName TEXT NOT NULL, UNIQUE(suiteName))");
     if (!getDatabase().tables().contains("TestsInSuite"))
         delete dbExec("CREATE TABLE TestsInSuite (id INTEGER PRIMARY KEY, testSuiteID INTEGER NOT NULL, testID INTEGER NOT NULL)");
-    // if (!getDatabase().tables().contains("TestArgs"))
-    //     dbExec("CREATE TABLE TestArg (id INTEGER PRIMARY KEY, testID INTEGER NOT NULL, argIdx INTEGER NOT NULL, arg TEXT NOT NULL, isVarArg BOOL NOT NULL, defaultVal TEXT)");
+    if (!getDatabase().tables().contains("TestArgs"))
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Changes TestArgs to Arg

File: src/gui/VerificationValidationWidget.cpp [hunk 9]
```diff
@@ -220,8 +267,8 @@ void VerificationValidationWidget::dbInitTables() {
         delete dbExec("CREATE TABLE TestSuites (id INTEGER PRIMARY KEY, suiteName TEXT NOT NULL, UNIQUE(suiteName))");
     if (!getDatabase().tables().contains("TestsInSuite"))
         delete dbExec("CREATE TABLE TestsInSuite (id INTEGER PRIMARY KEY, testSuiteID INTEGER NOT NULL, testID INTEGER NOT NULL)");
-    // if (!getDatabase().tables().contains("TestArgs"))
-    //     dbExec("CREATE TABLE TestArg (id INTEGER PRIMARY KEY, testID INTEGER NOT NULL, argIdx INTEGER NOT NULL, arg TEXT NOT NULL, isVarArg BOOL NOT NULL, defaultVal TEXT)");
+    if (!getDatabase().tables().contains("TestArgs"))
+        dbExec("CREATE TABLE TestArg (id INTEGER PRIMARY KEY, testID INTEGER NOT NULL, argIdx INTEGER NOT NULL, arg TEXT NOT NULL, isVarArg BOOL NOT NULL, defaultVal TEXT)");
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Add delete

File: src/gui/VerificationValidationWidget.cpp [hunk 10]
```diff
@@ -256,16 +303,28 @@ void VerificationValidationWidget::dbPopulateDefaults() {
...
         for (int i = 0; i < DefaultTests::allTests.size(); i++) {
-            q->prepare("INSERT INTO Tests (testName, testCommand) VALUES (?, ?)");
-            q->addBindValue(DefaultTests::allTests[i].testName);
-            q->addBindValue(DefaultTests::allTests[i].testCommand);
+            q->prepare("INSERT INTO Tests (testName, testCommand, hasValArgs, category) VALUES (:testName, :testCommand, :hasValArgs, :category)");
+            q->bindValue(":testName", DefaultTests::allTests[i].testName);
+            q->bindValue(":testCommand", DefaultTests::allTests[i].testCommand);
+            q->bindValue(":hasValArgs", DefaultTests::allTests[i].hasVariable);
+            q->bindValue(":category", DefaultTests::allTests[i].category);
             dbExec(q);
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:311]: Add delete

File: src/gui/VerificationValidationWidget.cpp [hunk 11]
```diff
@@ -286,11 +345,31 @@ void VerificationValidationWidget::dbPopulateDefaults() {
...
+            item = testList->item(i);
+            if(!tests.contains(item) || item->toolTip() == "category")
+                item->setHidden(true);
+            else
+                item->setHidden(false);
+        }
+    }
+}
+
+void VerificationValidationWidget::searchSuites(const QString &input)  {
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:450]: Might not be necessary

File: src/gui/VerificationValidationWidget.cpp [hunk 12]
```diff
@@ -378,13 +463,148 @@ void VerificationValidationWidget::updateTestListWidget(QListWidgetItem* suite_c
...
+    if (q->first()) {
+        if(test_clicked->checkState()) {
+            selectedTests << q->value(0).toString();
+        }
+        else {
+            int index = selectedTests.indexOf(QRegExp(q->value(0).toString()));
+            selectedTests.removeAt(index);
+        }
+    }
+
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: delete q

File: src/gui/VerificationValidationWidget.cpp [hunk 13]
```diff
@@ -378,13 +463,148 @@ void VerificationValidationWidget::updateTestListWidget(QListWidgetItem* suite_c
...
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:578]: delete q

File: src/gui/VerificationValidationWidget.cpp [hunk 14]
```diff
@@ -378,13 +463,148 @@ void VerificationValidationWidget::updateTestListWidget(QListWidgetItem* suite_c
...
+        q->prepare("INSERT INTO TestsInSuite (testSuiteID, testID) VALUES (?, ?)");
+        q->addBindValue(testSuiteID);
+        q->addBindValue(testID);
+        dbExec(q);  
+    }
+
+    ///testList->clear();
+    ///suiteList->clear();
+    ///selectedTests.clear();
+    newTestSuiteDialog->close();
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:599]: delete q

File: src/gui/VerificationValidationWidget.cpp [hunk 15]
```diff
@@ -378,13 +463,148 @@ void VerificationValidationWidget::updateTestListWidget(QListWidgetItem* suite_c
...
+        q->prepare("DELETE FROM TestSuites WHERE id = :suiteID");
+        q->bindValue(":suiteID", suite);
+        dbExec(q,!SHOW_ERROR_POPUP); 
+
+        q->prepare("DELETE FROM TestsInSuite WHERE testSuiteID = :suiteID");
+        q->bindValue(":suiteID", suite);
+        dbExec(q,!SHOW_ERROR_POPUP); 
+    }
+
+    removeTestSuiteDialog->close();
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:617]: delete q

File: src/gui/VerificationValidationWidget.cpp [hunk 16]
```diff
@@ -378,13 +463,148 @@ void VerificationValidationWidget::updateTestListWidget(QListWidgetItem* suite_c
...
+    QSqlQuery* q = new QSqlQuery(getDatabase());
+
+    bool isVariable = false;
+    for (int i = 0; i < groupBoxVector.size(); i++) {
+        QList<QCheckBox*> checkboxList = groupBoxVector.at(i)->findChildren<QCheckBox*>();
+        if (checkboxList.at(0)->isChecked())
+            isVariable = true;
+    }
+
+    QList<QLineEdit*> testInfoList = groupbox1->findChildren<QLineEdit*>();
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Consider renaming groupbox1

File: src/gui/VerificationValidationWidget.cpp [hunk 17]
```diff
@@ -378,13 +463,148 @@ void VerificationValidationWidget::updateTestListWidget(QListWidgetItem* suite_c
...
+                q->addBindValue(testID);
+                q->addBindValue(i);
+                q->addBindValue(arg);
+                q->addBindValue(isVariable);
+            }
+
+            dbExec(q); 
+        }
+    }
+
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:668]: delete q

File: src/gui/VerificationValidationWidget.cpp [hunk 18]
```diff
@@ -418,33 +640,436 @@ void VerificationValidationWidget::testListSelection(QListWidgetItem* test_click
...
+        q->prepare("DELETE FROM TestsInSuite WHERE testID = :testID");
+        q->bindValue(":testID", test);
+        dbExec(q,!SHOW_ERROR_POPUP); 
+
+        // uncommennt when the database is update to have args
+        q->prepare("DELETE FROM TestArg WHERE testID = :testID");
+        q->bindValue(":testID", test);
+        dbExec(q,!SHOW_ERROR_POPUP);
+    }
+
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:734]: delete q

File: src/gui/VerificationValidationWidget.cpp [hunk 19]
```diff
@@ -418,33 +640,436 @@ void VerificationValidationWidget::testListSelection(QListWidgetItem* test_click
...
+    removeTestDialog = new QDialog();
+    removeTestDialog->setModal(true);
+    removeTestDialog->setWindowTitle("Remove test");
 
-    QLineEdit* input1 = new QLineEdit();
-    QLineEdit* input2 = new QLineEdit();
+    // Get test list from db
+    QSqlDatabase db = getDatabase();
+    QSqlQuery query(db);
+    query.exec("Select testName, testCommand from Tests ORDER BY id ASC");
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Inconsistent SQL commands, all caps

File: src/gui/VerificationValidationWidget.cpp [hunk 20]
```diff
@@ -418,33 +640,436 @@ void VerificationValidationWidget::testListSelection(QListWidgetItem* test_click
...
+    searchBar->addWidget(searchBox);
+	
+    // format and populate Select Tests dialog box
+    QGridLayout* grid = new QGridLayout();
+
+    
+    // Get suite list from db
+    QSqlDatabase db = getDatabase();
+    QSqlQuery query(db);
+    query.exec("Select suiteName from TestSuites ORDER by id ASC");
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Inconsistent SQL commands, all caps

File: src/gui/VerificationValidationWidget.cpp [hunk 21]
```diff
@@ -418,33 +640,436 @@ void VerificationValidationWidget::testListSelection(QListWidgetItem* test_click
...
+    selectedTests.clear();
+
+    newTestSuiteDialog = new QDialog();
+    newTestSuiteDialog->setModal(true);
+    newTestSuiteDialog->setWindowTitle("Create new test suite");
+
+    // Get test list from db
+    QSqlDatabase db = getDatabase();
+    QSqlQuery query(db);
+    query.exec("Select testName, testCommand from Tests ORDER BY id ASC");
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp]: Inconsistent SQL commands, all caps

File: src/gui/VerificationValidationWidget.cpp [hunk 22]
```diff
@@ -418,33 +640,436 @@ void VerificationValidationWidget::testListSelection(QListWidgetItem* test_click
...
+
+
+// void VerificationValidationWidget::updateVarArgs(QString testName, int index, QString input){
+//     std::cout << "called"  << std::endl;
+//     std::cout << testName.toStdString() << std::endl;
+//     std::cout << index << std::endl;
+//     std::cout << input.toStdString() << std::endl;
+// }
+
+void VerificationValidationWidget::updateVarArgs() {
```
Comment (daperego) [inline src/gui/VerificationValidationWidget.cpp:1098]: Check if this is necessary

**Review by daperego — CHANGES_REQUESTED** [review]
Reference in-line comments

File: src/gui/VerificationValidationWidget.cpp (hunk 4 above)
Comment (gJubey) [reply src/gui/VerificationValidationWidget.cpp]: moving this to the .h resulted in an error im not sure the cause of so i left it were it was

**Pull request closed without merging**